src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/NumericIndexWriter.java

```diff
@@ -0,0 +1,179 @@
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.v1.segment.SegmentMetadata;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.packed.PackedInts;
+import org.apache.lucene.util.packed.PackedLongValues;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+
+/**
+ * Specialized writer for 1-dim point values, that builds them into a {@link BlockBalancedTreeWriter} with auxiliary
+ * posting lists on eligible tree levels.
+ *
+ * Given a sorted input {@link IntersectingPointValues}, the flush process is optimised because we don't need to
+ * buffer all point values to sort them.
+ */
+public class NumericIndexWriter
+{
+    public static final int MAX_POINTS_IN_LEAF_NODE = BlockBalancedTreeWriter.DEFAULT_MAX_POINTS_IN_LEAF_NODE;
+
+    private final BlockBalancedTreeWriter writer;
+    private final IndexDescriptor indexDescriptor;
+    private final IndexContext indexContext;
+    private final int bytesPerValue;
+
+    /**
+     * @param maxSegmentRowId maximum possible segment row ID, used to create `maxDoc` for the balanced tree
+     */
+    public NumericIndexWriter(IndexDescriptor indexDescriptor,
+                              IndexContext indexContext,
+                              int bytesPerValue,
+                              long maxSegmentRowId)
+    {
+        this(indexDescriptor, indexContext, MAX_POINTS_IN_LEAF_NODE, bytesPerValue, maxSegmentRowId);
+    }
+
+    @VisibleForTesting
+    public NumericIndexWriter(IndexDescriptor indexDescriptor,
+                              IndexContext indexContext,
+                              int maxPointsInLeafNode,
+                              int bytesPerValue,
+                              long maxSegmentRowId)
+    {
+        checkArgument(maxSegmentRowId >= 0, "[%s] maxRowId must be non-negative value, but got %s", indexContext.getIndexName(), maxSegmentRowId);
+
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.bytesPerValue = bytesPerValue;
+        this.writer = new BlockBalancedTreeWriter(maxSegmentRowId + 1, bytesPerValue, maxPointsInLeafNode);
+    }
+
+    @Override
+    public String toString()
+    {
+        return MoreObjects.toStringHelper(this)
+                          .add("bytesPerDim", bytesPerValue)
+                          .add("bufferedPoints", writer.getPointCount())
+                          .toString();
+    }
+
+    public static class LeafCallback implements BlockBalancedTreeWriter.Callback
+    {
+        final List<PackedLongValues> postings = new ArrayList<>();
```

Comment:
nit: Looking for small optimizations here. Is there any way we could estimate the number of leaves when we create this list rather than potentially having to grow a few times? I guess this is at the per-segment level, so not super important.

Reply:
Yeah. I guess it doesn't have to be a super accurate estimate...just enough to avoid silly `grow()` calls.